src/lir/lir_type.rs

```diff
@@ -0,0 +1,157 @@
+        returns: Vec<Type>,
+    },
+    Mapping {
+        key_ty: Box<Type>,
+        value_ty: Box<Type>,
+    },
+    Array(Box<Type>, Vec<ArrayLength>),
+    Struct(StructType),
+    Slice(Box<Type>),
+}
+
+#[derive(Clone, Debug)]
+pub enum InternalCallTy {
+    Static { cfg_no: usize },
+    Dynamic(Operand),
+    Builtin { ast_func_no: usize },
+}
+
+#[derive(Clone, Debug)]
+pub struct PhiInput {
+    pub operand: Operand,
+    pub block_no: usize,
+}
+
+impl From<&ast::StructType> for StructType {
+    fn from(ty: &ast::StructType) -> Self {
+        match ty {
+            ast::StructType::AccountInfo => StructType::SolAccountInfo,
+            ast::StructType::AccountMeta => StructType::SolAccountMeta,
+            ast::StructType::ExternalFunction => StructType::ExternalFunction,
+            ast::StructType::SolParameters => StructType::SolParameters,
+            ast::StructType::UserDefined(i) => StructType::UserDefined(*i),
+        }
+    }
+}
+
+impl fmt::Display for StructType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            StructType::UserDefined(i) => write!(f, "{}", i),
+            StructType::SolAccountInfo => write!(f, "SolAccountInfo"),
+            StructType::SolAccountMeta => write!(f, "SolAccountMeta"),
+            StructType::ExternalFunction => write!(f, "ExternalFunction"),
+            StructType::SolParameters => write!(f, "SolParameters"),
+            StructType::Vector(elem_ty) => write!(f, "vector<{}>", elem_ty),
+        }
+    }
+}
+
+impl fmt::Display for LIRType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.lir_type)
+    }
+}
+
+impl fmt::Display for Type {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Type::Bool => write!(f, "bool"),
+            Type::Int(width) => write!(f, "int{}", width),
+            Type::Uint(width) => write!(f, "uint{}", width),
+            Type::Bytes(width) => write!(f, "bytes{}", width),
+            Type::Ptr(ty) => write!(f, "ptr<{}>", ty),
+            Type::StoragePtr(immutable, ty) => {
+                if *immutable {
+                    write!(f, "const_storage_ptr<{}>", ty)
+                } else {
+                    write!(f, "storage_ptr<{}>", ty)
+                }
+            }
+            Type::Array(ty, len) => {
+                write!(f, "{}", ty)?;
+                len.iter().for_each(|len| match len {
+                    ArrayLength::Fixed(len) => write!(f, "[{}]", len).unwrap(),
+                    ArrayLength::Dynamic => write!(f, "[]").unwrap(),
+                    ArrayLength::AnyFixed => write!(f, "[?]").unwrap(),
+                });
+                Ok(())
+            }
+            Type::Slice(ty) => write!(f, "slice<{}>", ty),
+            Type::Struct(ty) => write!(f, "struct.{}", ty),
+            Type::Function { params, returns } => {
+                write!(f, "fn(")?;
```

Comment:
Your still inventing new syntax for Solidity types here.

Reply:
Hi @seanyoung, 
I updated the function type printing, now it's identical to Solidity syntax.

For example:
```rust
print!("{}", lir_type::Type::Function {
    params: vec![Type::Uint(8), Type::Uint(64)],
    returns: vec![Type::Uint(8), Type::Uint(8)],
});
```
The above code will print:
`function (uint8, uint64) returns (uint8, uint8)`

Reference: https://github.com/ethereum/solidity/issues/4349